File: tests/domains/video/test_narration.py
import subprocess

import pytest

from core.paths import Workspace
from domains import Runtime
from domains.video.narration import add_narration
from domains.video.probe import probe_video
from tests.helpers import unwrap

pytestmark = pytest.mark.ffmpeg


@pytest.fixture
def narration(workspace: Workspace) -> str:
    target = workspace.root / "voz.m4a"
    subprocess.run(
        [
            "ffmpeg",
            "-hide_banner",
            "-loglevel",
            "error",
            "-y",
            "-f",
            "lavfi",
            "-i",
            "sine=frequency=880:duration=1.5",
            "-c:a",
            "aac",
            str(target),
        ],
        check=True,
    )
    return "voz.m4a"


def test_add_narration_mix(runtime: Runtime, sample_video: str, narration: str) -> None:
    result = unwrap(add_narration(runtime, sample_video, narration, start=0.5))
    assert result["output"] == "sample_narrated.mp4"
    assert result["replaced_audio"] is False
    probed = unwrap(probe_video(runtime, result["output"]))
    assert probed["info"]["has_audio"]
    assert abs(probed["info"]["duration"] - 3.0) < 0.3


def test_add_narration_replace(runtime: Runtime, sample_video: str, narration: str) -> None:
    result = unwrap(add_narration(runtime, sample_video, narration, original_volume=0.0))
    assert result["replaced_audio"] is True


def test_add_narration_invalid(runtime: Runtime, sample_video: str, narration: str) -> None:
    assert add_narration(runtime, sample_video, narration, start=-1).get("code") == (
        "invalid_argument"
    )
    assert add_narration(runtime, sample_video, narration, original_volume=2).get("code") == (
        "invalid_argument"
    )
    assert add_narration(runtime, sample_video, narration, start=50).get("code") == (
        "invalid_argument"
    )


def test_add_narration_missing(runtime: Runtime, sample_video: str) -> None:
    assert add_narration(runtime, sample_video, "ghost.mp3").get("code") == "not_found"

File: tests/domains/video/test_narration.rs
use std::process::Command;

use mcp_tools::core::errors::ErrorCode;
use mcp_tools::core::paths::Workspace;
use mcp_tools::domains::video::narration::add_narration;
use mcp_tools::domains::video::probe::probe_video;

use crate::conftest::{runtime, sample_video};
use crate::helpers::unwrap;
use crate::skip_without_ffmpeg;

/// Gera 1,5s de narração sintética (`voz.m4a`) no workspace.
fn narration(workspace: &Workspace) -> String {
    let target = workspace.root.join("voz.m4a");
    let status = Command::new("ffmpeg")
        .args([
            "-hide_banner",
            "-loglevel",
            "error",
            "-y",
            "-f",
            "lavfi",
            "-i",
            "sine=frequency=880:duration=1.5",
            "-c:a",
            "aac",
        ])
        .arg(&target)
        .status()
        .expect("ffmpeg");
    assert!(status.success());
    "voz.m4a".to_string()
}

#[test]
fn test_add_narration_mix() {
    skip_without_ffmpeg!();
    let rt = runtime();
    let sample = sample_video(&rt.runtime.workspace);
    let voice = narration(&rt.runtime.workspace);
    let result = unwrap(add_narration(&rt.runtime, &sample, &voice, 0.5, 0.2, false));
    assert_eq!(result.output, "sample_narrated.mp4");
    assert!(!result.replaced_audio);
    let probed = probe_video(&rt.runtime, &result.output).unwrap();
    assert!(probed.info.has_audio);
    assert!((probed.info.duration - 3.0).abs() < 0.3);
}

#[test]
fn test_add_narration_replace() {
    skip_without_ffmpeg!();
    let rt = runtime();
    let sample = sample_video(&rt.runtime.workspace);
    let voice = narration(&rt.runtime.workspace);
    let result = unwrap(add_narration(&rt.runtime, &sample, &voice, 0.0, 0.0, false));
    assert!(result.replaced_audio);
}

#[test]
fn test_add_narration_invalid() {
    skip_without_ffmpeg!();
    let rt = runtime();
    let sample = sample_video(&rt.runtime.workspace);
    let voice = narration(&rt.runtime.workspace);
    let error = add_narration(&rt.runtime, &sample, &voice, -1.0, 0.2, false).unwrap_err();
    assert_eq!(error.code, ErrorCode::InvalidArgument);
    let error = add_narration(&rt.runtime, &sample, &voice, 0.0, 2.0, false).unwrap_err();
    assert_eq!(error.code, ErrorCode::InvalidArgument);
    let error = add_narration(&rt.runtime, &sample, &voice, 50.0, 0.2, false).unwrap_err();
    assert_eq!(error.code, ErrorCode::InvalidArgument);
}

#[test]
fn test_add_narration_missing() {
    skip_without_ffmpeg!();
    let rt = runtime();
    let sample = sample_video(&rt.runtime.workspace);
    let error = add_narration(&rt.runtime, &sample, "ghost.mp3", 0.0, 0.2, false).unwrap_err();
    assert_eq!(error.code, ErrorCode::NotFound);
}
